# DNAvigators

## Abstract

### Motivation:

An abundance of genomic information exists in modes such as RNA-seq, ATAC-seq, and CITE-seq. Computational biologists have taken advantage of this information with the application of deep learning methods like convolutional neural networks (CNNs) and generative approaches (GANs, autoencoders). Recently, several state-of-the-art networks have obtained desirable results on single-cell analysis of COVID-19 disease factors through the alignment of relevant datasets in a latent space. These spaces are formed without a specific task in mind; we aim to incorporate cell-type specific information to improve the utility of the outputted space for tasks such as classification of case severity.

### Results:

By guiding alignment of data from different datasets with the use of cell-type-specific information, CHEESgrate may be used to attain better classification results than comparable models with inputs from unguided training. Further, they offer improved interpretation of contributions from cell-type-specific factors, providing biologically valuable information.


##### Data:

Data utilized in training may be found within their respective training files.

Contact: alexander_halpin@brown.edu, brendan_leahey@brown.edu, micah_lessnick@brown.edu, winston_li@brown.edu
